File: sdk/go/goodkey/cert/cert.go
// Package cert provides X.509 certificate validation utilities for CEF
// implementations that use certificate-backed recipients.
//
// This is part of the GoodKey implementation layer, not the format layer.
// The CEF format itself is certificate-agnostic — it identifies recipients
// by key ID. This package provides the validation logic for backends that
// resolve key IDs to certificates.
package cert

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"fmt"
	"time"
)

// Validator validates certificates for use in CEF operations.
type Validator interface {
	ValidateForEncryption(cert *x509.Certificate) error
	ValidateForSigning(cert *x509.Certificate) error
	ValidateChain(cert *x509.Certificate, intermediates []*x509.Certificate) error
}

// StandardValidator implements standard certificate validation with
// configurable checks for expiry, key usage, and self-signed acceptance.
type StandardValidator struct {
	CheckExpiry      bool
	CheckKeyUsage    bool
	AllowExpired     bool
	AllowSelfSigned  bool

	// nowFn can be overridden for testing. Defaults to time.Now.
	nowFn func() time.Time
}

// DefaultValidator is the default certificate validator.
var DefaultValidator Validator = &StandardValidator{
	CheckExpiry:   true,
	CheckKeyUsage: true,
}

// now returns the current time, using nowFn if set.
func (v *StandardValidator) now() time.Time {
	if v.nowFn != nil {
		return v.nowFn()
	}
	return time.Now()
}

// ValidateForEncryption checks if a certificate is suitable for encrypting to.
func (v *StandardValidator) ValidateForEncryption(cert *x509.Certificate) error {
	if cert == nil {
		return fmt.Errorf("certificate is nil")
	}

	if v.CheckExpiry && !v.AllowExpired {
		now := v.now()
		if now.Before(cert.NotBefore) {
			return &ValidationError{
				Certificate: cert,
				Reason:      fmt.Sprintf("certificate not yet valid (notBefore: %s)", cert.NotBefore.Format(time.RFC3339)),
			}
		}
		if now.After(cert.NotAfter) {
			return &ValidationError{
				Certificate: cert,
				Reason:      fmt.Sprintf("certificate expired (notAfter: %s)", cert.NotAfter.Format(time.RFC3339)),
			}
		}
	}

	if v.CheckKeyUsage {
		// For encryption recipients, the certificate needs KeyEncipherment (RSA)
		// or KeyAgreement (ECDH). Some certs have both.
		hasEncryptionUsage := cert.KeyUsage&x509.KeyUsageKeyEncipherment != 0 ||
			cert.KeyUsage&x509.KeyUsageKeyAgreement != 0
		// If KeyUsage is zero, the extension is absent — allow by default
		if cert.KeyUsage != 0 && !hasEncryptionUsage {
			return &ValidationError{
				Certificate: cert,
				Reason:      "certificate does not have KeyEncipherment or KeyAgreement usage",
			}
		}
	}

	return nil
}

// ValidateForSigning checks if a certificate is suitable for signature verification.
func (v *StandardValidator) ValidateForSigning(cert *x509.Certificate) error {
	if cert == nil {
		return fmt.Errorf("certificate is nil")
	}

	if v.CheckExpiry && !v.AllowExpired {
		now := v.now()
		if now.Before(cert.NotBefore) {
			return &ValidationError{
				Certificate: cert,
				Reason:      fmt.Sprintf("certificate not yet valid (notBefore: %s)", cert.NotBefore.Format(time.RFC3339)),
			}
		}
		if now.After(cert.NotAfter) {
			return &ValidationError{
				Certificate: cert,
				Reason:      fmt.Sprintf("certificate expired (notAfter: %s)", cert.NotAfter.Format(time.RFC3339)),
			}
		}
	}

	if v.CheckKeyUsage {
		hasSigningUsage := cert.KeyUsage&x509.KeyUsageDigitalSignature != 0 ||
			cert.KeyUsage&x509.KeyUsageContentCommitment != 0
		if cert.KeyUsage != 0 && !hasSigningUsage {
			return &ValidationError{
				Certificate: cert,
				Reason:      "certificate does not have DigitalSignature usage",
			}
		}
	}

	return nil
}

// ValidateChain validates a certificate against system roots with optional intermediates.
func (v *StandardValidator) ValidateChain(cert *x509.Certificate, intermediates []*x509.Certificate) error {
	if cert == nil {
		return fmt.Errorf("certificate is nil")
	}

	if v.AllowSelfSigned && cert.IsCA {
		return nil
	}

	pool := x509.NewCertPool()
	for _, ic := range intermediates {
		pool.AddCert(ic)
	}

	opts := x509.VerifyOptions{
		Intermediates: pool,
		CurrentTime:   v.now(),
	}

	if _, err := cert.Verify(opts); err != nil {
		return &ValidationError{
			Certificate: cert,
			Reason:      fmt.Sprintf("chain validation failed: %v", err),
		}
	}

	return nil
}

// NoOpValidator accepts all certificates without validation.
// Useful for testing and development environments.
type NoOpValidator struct{}

func (v *NoOpValidator) ValidateForEncryption(*x509.Certificate) error { return nil }
func (v *NoOpValidator) ValidateForSigning(*x509.Certificate) error    { return nil }
func (v *NoOpValidator) ValidateChain(*x509.Certificate, []*x509.Certificate) error {
	return nil
}

// ValidationError represents a certificate validation failure.
type ValidationError struct {
	Certificate *x509.Certificate
	Reason      string
}

func (e *ValidationError) Error() string {
	if e.Certificate != nil {
		return fmt.Sprintf("certificate validation failed for %s: %s", e.Certificate.Subject.CommonName, e.Reason)
	}
	return fmt.Sprintf("certificate validation failed: %s", e.Reason)
}

// ---------------------------------------------------------------------------
// PEM utilities
// ---------------------------------------------------------------------------

// ParsePEM parses a single PEM-encoded certificate.
func ParsePEM(data []byte) (*x509.Certificate, error) {
	block, _ := pem.Decode(data)
	if block == nil {
		return nil, fmt.Errorf("no PEM block found")
	}
	if block.Type != "CERTIFICATE" {
		return nil, fmt.Errorf("PEM block type is %q, expected CERTIFICATE", block.Type)
	}
	return x509.ParseCertificate(block.Bytes)
}

// ParsePEMChain parses multiple PEM-encoded certificates from a single byte slice.
func ParsePEMChain(data []byte) ([]*x509.Certificate, error) {
	var certs []*x509.Certificate
	rest := data
	for {
		var block *pem.Block
		block, rest = pem.Decode(rest)
		if block == nil {
			break
		}
		if block.Type != "CERTIFICATE" {
			continue
		}
		cert, err := x509.ParseCertificate(block.Bytes)
		if err != nil {
			return nil, fmt.Errorf("parse certificate: %w", err)
		}
		certs = append(certs, cert)
	}
	if len(certs) == 0 {
		return nil, fmt.Errorf("no certificates found in PEM data")
	}
	return certs, nil
}

// MatchIssuerAndSerial checks if a certificate matches the given issuer DN and serial number.
func MatchIssuerAndSerial(cert *x509.Certificate, issuer []byte, serial []byte) bool {
	if cert == nil {
		return false
	}
	if !bytes.Equal(cert.RawIssuer, issuer) {
		return false
	}
	return bytes.Equal(cert.SerialNumber.Bytes(), serial)
}

// KeyType returns a human-readable description of the certificate's public key type.
func KeyType(cert *x509.Certificate) string {
	switch cert.PublicKey.(type) {
	case *rsa.PublicKey:
		pk := cert.PublicKey.(*rsa.PublicKey)
		return fmt.Sprintf("RSA-%d", pk.N.BitLen())
	case *ecdsa.PublicKey:
		pk := cert.PublicKey.(*ecdsa.PublicKey)
		return fmt.Sprintf("ECDSA-%s", pk.Curve.Params().Name)
	case ed25519.PublicKey:
		return "Ed25519"
	default:
		return fmt.Sprintf("%T", cert.PublicKey)
	}
}

File: sdk/typescript/src/x509/cert.ts
/**
 * Certificate validation utilities for CEF implementations that use
 * certificate-backed recipients.
 *
 * Uses @peculiar/x509 for real X.509 parsing, key usage checking,
 * and chain validation. Works in browsers and Node.js.
 *
 * This module mirrors sdk/go/goodkey/cert/cert.go.
 */

import 'reflect-metadata';
import * as x509 from '@peculiar/x509';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface CertificateInfo {
  /** The parsed @peculiar/x509 certificate. */
  cert: x509.X509Certificate;

  /** Subject string (e.g., "CN=Alice"). */
  subject: string;

  /** Issuer string (e.g., "CN=GoodKey CA"). */
  issuer: string;

  /** Serial number (hex-encoded). */
  serialNumber: string;

  /** Subject Key Identifier (hex-encoded, if present). */
  subjectKeyIdentifier?: string;

  /** Not valid before. */
  notBefore: Date;

  /** Not valid after. */
  notAfter: Date;

  /** Public key algorithm name. */
  publicKeyAlgorithm: string;

  /** DER-encoded certificate bytes. */
  raw: Uint8Array;
}

export interface ValidatorOptions {
  checkExpiry?: boolean;
  checkKeyUsage?: boolean;
  allowExpired?: boolean;
}

// ---------------------------------------------------------------------------
// Validator interface
// ---------------------------------------------------------------------------

export interface CertificateValidator {
  validateForEncryption(cert: CertificateInfo): void;
  validateForSigning(cert: CertificateInfo): void;
  validateChain(cert: CertificateInfo, intermediates?: CertificateInfo[], roots?: x509.X509Certificate[]): Promise<void>;
}

// ---------------------------------------------------------------------------
// StandardValidator
// ---------------------------------------------------------------------------

export class StandardValidator implements CertificateValidator {
  private checkExpiry: boolean;
  private checkKeyUsage: boolean;
  private allowExpired: boolean;
  private nowFn: () => Date;

  constructor(opts?: ValidatorOptions & { nowFn?: () => Date }) {
    this.checkExpiry = opts?.checkExpiry ?? true;
    this.checkKeyUsage = opts?.checkKeyUsage ?? true;
    this.allowExpired = opts?.allowExpired ?? false;
    this.nowFn = opts?.nowFn ?? (() => new Date());
  }

  validateForEncryption(info: CertificateInfo): void {
    this.checkExpiryIfEnabled(info);

    if (this.checkKeyUsage) {
      const kuExt = info.cert.getExtension(x509.KeyUsagesExtension);
      if (kuExt) {
        const hasKeyEncipherment = !!(kuExt.usages & x509.KeyUsageFlags.keyEncipherment);
        const hasKeyAgreement = !!(kuExt.usages & x509.KeyUsageFlags.keyAgreement);
        if (!hasKeyEncipherment && !hasKeyAgreement) {
          throw new Error(
            `certificate "${info.subject}" does not have keyEncipherment or keyAgreement usage`,
          );
        }
      }
    }
  }

  validateForSigning(info: CertificateInfo): void {
    this.checkExpiryIfEnabled(info);

    if (this.checkKeyUsage) {
      const kuExt = info.cert.getExtension(x509.KeyUsagesExtension);
      if (kuExt) {
        const hasDigitalSignature = !!(kuExt.usages & x509.KeyUsageFlags.digitalSignature);
        if (!hasDigitalSignature) {
          throw new Error(
            `certificate "${info.subject}" does not have digitalSignature usage`,
          );
        }
      }
    }
  }

  async validateChain(
    info: CertificateInfo,
    intermediates?: CertificateInfo[],
    roots?: x509.X509Certificate[],
  ): Promise<void> {
    this.checkExpiryIfEnabled(info);

    if (roots && roots.length > 0) {
      const chainBuilder = new x509.X509ChainBuilder({
        certificates: intermediates?.map(i => i.cert) ?? [],
      });

      try {
        const chain = await chainBuilder.build(info.cert);
        // Verify the chain ends at a trusted root
        const rootCert = chain[chain.length - 1];
        if (rootCert) {
          const rootMatch = roots.some(r =>
            r.serialNumber === rootCert.serialNumber &&
            r.subject === rootCert.subject
          );
          if (!rootMatch) {
            throw new Error('chain does not terminate at a trusted root');
          }
        }
      } catch (e) {
        throw new Error(`chain verification failed: ${(e as Error).message}`);
      }
    }
  }

  private checkExpiryIfEnabled(info: CertificateInfo): void {
    if (!this.checkExpiry || this.allowExpired) return;

    const now = this.nowFn();
    if (now < info.notBefore) {
      throw new Error(
        `certificate "${info.subject}" not yet valid (notBefore: ${info.notBefore.toISOString()})`,
      );
    }
    if (now > info.notAfter) {
      throw new Error(
        `certificate "${info.subject}" expired (notAfter: ${info.notAfter.toISOString()})`,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// NoOpValidator (testing only)
// ---------------------------------------------------------------------------

export class NoOpValidator implements CertificateValidator {
  validateForEncryption(_info: CertificateInfo): void {}
  validateForSigning(_info: CertificateInfo): void {}
  async validateChain(_info: CertificateInfo): Promise<void> {}
}

// ---------------------------------------------------------------------------
// Certificate parsing
// ---------------------------------------------------------------------------

/**
 * Parse a DER-encoded certificate into CertificateInfo.
 */
export function parseCertificateDER(der: Uint8Array): CertificateInfo {
  const cert = new x509.X509Certificate(der as unknown as BufferSource);
  return certToInfo(cert);
}

/**
 * Parse a PEM-encoded certificate into CertificateInfo.
 */
export function parseCertificatePEM(pem: string): CertificateInfo {
  const cert = new x509.X509Certificate(pem);
  return certToInfo(cert);
}

/**
 * Parse a PEM chain (multiple certificates) into CertificateInfo[].
 */
export function parseCertificateChainPEM(pem: string): CertificateInfo[] {
  const certs = x509.PemConverter.decode(pem);
  return certs.map(der => {
    const cert = new x509.X509Certificate(der);
    return certToInfo(cert);
  });
}

function certToInfo(cert: x509.X509Certificate): CertificateInfo {
  const skiExt = cert.getExtension(x509.SubjectKeyIdentifierExtension);

  return {
    cert,
    subject: cert.subject,
    issuer: cert.issuer,
    serialNumber: cert.serialNumber,
    subjectKeyIdentifier: skiExt?.keyId,
    notBefore: cert.notBefore,
    notAfter: cert.notAfter,
    publicKeyAlgorithm: cert.publicKey.algorithm.name,
    raw: new Uint8Array(cert.rawData),
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Match a certificate by issuer and serial number.
 */
export function matchIssuerAndSerial(
  info: CertificateInfo,
  issuer: string,
  serialNumber: string,
): boolean {
  return info.issuer === issuer && info.serialNumber === serialNumber;
}

/**
 * Create a default StandardValidator with recommended settings.
 */
export function defaultValidator(): StandardValidator {
  return new StandardValidator();
}

/**
 * Generate a self-signed test certificate.
 * Requires crypto provider to be set (Node.js: set in test setup).
 */
export async function generateTestCertificate(opts: {
  subject: string;
  keyUsages?: number;
  notBefore?: Date;
  notAfter?: Date;
}): Promise<CertificateInfo> {
  const alg = { name: 'ECDSA', namedCurve: 'P-256', hash: 'SHA-256' };
  const keys = await globalThis.crypto.subtle.generateKey(
    alg as EcKeyGenParams,
    true,
    ['sign', 'verify'],
  );

  const extensions: x509.Extension[] = [];
  if (opts.keyUsages !== undefined) {
    extensions.push(new x509.KeyUsagesExtension(opts.keyUsages, true));
  }

  const cert = await x509.X509CertificateGenerator.createSelfSigned({
    serialNumber: Math.floor(Math.random() * 1000000).toString(16),
    name: `CN=${opts.subject}`,
    notBefore: opts.notBefore ?? new Date(),
    notAfter: opts.notAfter ?? new Date(Date.now() + 365 * 24 * 60 * 60 * 1000),
    keys: keys as CryptoKeyPair,
    signingAlgorithm: alg,
    extensions,
  });

  return certToInfo(cert);
}

/** Re-export KeyUsageFlags for convenience. */
export const KeyUsageFlags = x509.KeyUsageFlags;

// ---------------------------------------------------------------------------
// Self-Signed ML-DSA-65 Certificate
// ---------------------------------------------------------------------------

import { AsnConvert } from '@peculiar/asn1-schema';
import * as x509Asn from '@peculiar/asn1-x509';

/** OID for ML-DSA-65 (id-ml-dsa-65 from NIST CSOR) */
const OID_ML_DSA_65 = '2.16.840.1.101.3.4.3.18';

/**
 * Create a self-signed X.509v3 certificate with ML-DSA-65.
 *
 * Builds a certificate using the provided ML-DSA-65 key pair from
 * @noble/post-quantum. The certificate is signed directly with the
 * raw ML-DSA sign function — no WebCrypto needed.
 *
 * @param publicKey  ML-DSA-65 public key (1952 bytes)
 * @param signFn     Raw sign function: (message) => signature
 * @param subject    Subject CN (e.g., "Demo Sender")
 * @returns Base64-encoded DER certificate (for PV certificate viewer)
 */
export function createSelfSignedMLDSACert(
  publicKey: Uint8Array,
  signFn: (msg: Uint8Array) => Uint8Array,
  subject: string,
): string {
  const tbs = new x509Asn.TBSCertificate();
  tbs.version = x509Asn.Version.v3;

  // Random serial number (positive)
  const serial = new Uint8Array(16);
  crypto.getRandomValues(serial);
  serial[0] &= 0x7f;
  tbs.serialNumber = serial.buffer as ArrayBuffer;

  // ML-DSA-65 signature algorithm
  tbs.signature = new x509Asn.AlgorithmIdentifier({ algorithm: OID_ML_DSA_65 });

  // Issuer = Subject (self-signed)
  const name = new x509Asn.Name();
  const rdn = new x509Asn.RelativeDistinguishedName();
  rdn.push(new x509Asn.AttributeTypeAndValue({
    type: '2.5.4.3', // CN
    value: new x509Asn.AttributeValue({ printableString: subject }),
  }));
  name.push(rdn);
  tbs.issuer = name;
  tbs.subject = name;

  // Validity: now → +1 year
  const now = new Date();
  const expiry = new Date(now);
  expiry.setFullYear(expiry.getFullYear() + 1);
  tbs.validity = new x509Asn.Validity({
    notBefore: { utcTime: now } as any,
    notAfter: { utcTime: expiry } as any,
  });

  // ML-DSA-65 public key
  tbs.subjectPublicKeyInfo = new x509Asn.SubjectPublicKeyInfo({
    algorithm: new x509Asn.AlgorithmIdentifier({ algorithm: OID_ML_DSA_65 }),
    subjectPublicKey: publicKey.buffer as ArrayBuffer,
  });

  // Sign the TBSCertificate
  const tbsDer = new Uint8Array(AsnConvert.serialize(tbs));
  const signature = signFn(tbsDer);

  // Assemble the Certificate
  const cert = new x509Asn.Certificate({
    tbsCertificate: tbs,
    signatureAlgorithm: new x509Asn.AlgorithmIdentifier({ algorithm: OID_ML_DSA_65 }),
    signatureValue: signature.buffer as ArrayBuffer,
  });

  const certDer = new Uint8Array(AsnConvert.serialize(cert));
  return btoa(String.fromCharCode(...certDer));
}
